README.rst
Canonical JSON
==============

.. image:: https://img.shields.io/pypi/v/canonicaljson.svg
    :target: https://pypi.python.org/pypi/canonicaljson/
    :alt: Latest Version

Features
--------

* Encodes objects and arrays as `RFC 7159`_ JSON.
* Sorts object keys so that you get the same result each time.
* Has no insignificant whitespace to make the output as small as possible.
* Escapes only the characters that must be escaped, U+0000 to U+0019 / U+0022 /
  U+0056, to keep the output as small as possible.
* Uses the shortest escape sequence for each escaped character.
* Encodes the JSON as UTF-8.
* Can be configured to encode custom types unknown to the stdlib JSON encoder.

Supports Python versions 3.7 and newer.

.. _`RFC 7159`: https://tools.ietf.org/html/rfc7159

Installing
----------

.. code:: bash

   pip install canonicaljson

Using
-----

To encode an object into the canonicaljson:

.. code:: python

    import canonicaljson
    assert canonicaljson.encode_canonical_json({}) == b'{}'

There's also an iterator version:

.. code:: python

    import canonicaljson
    assert b''.join(canonicaljson.iterencode_canonical_json({})) == b'{}'

A preserialisation hook allows you to encode objects which aren't encodable by the
standard library ``JSONEncoder``.

.. code:: python

    import canonicaljson
    from typing import Dict

    class CustomType:
        pass

    def callback(c: CustomType) -> Dict[str, str]:
        return {"Hello": "world!"}

    canonicaljson.register_preserialisation_callback(CustomType, callback)
    assert canonicaljson.encode_canonical_json(CustomType()) == b'{"Hello":"world!"}'


CHANGES.md
Version 2.0.0 released 2023-03-15

Additions:

* Add a generic `register_preserialisation_callback` mechanism, which
  allows users to teach canonicaljson how to JSON-encode custom types.

Breaking changes:

* Remove support for serialising `frozendict` instances. Use the new
  `register_preserialisation_callback` mechanism to replace this
  functionality if needed.
* Remove support for `simplejson` and the `set_json_library`alternative
  json libraries.

Version 1.6.5 released 2023-02-15

* Update type hints to pass under mypy 1.0.

Version 1.6.4 released 2022-11-04

* Remove unused `setuptools_scm` build requirement.

Version 1.6.3 released 2022-09-23

* Properly package the `canonicaljson` module.

Version 1.6.2 released 2022-06-08

* Bump version to 1.6.2

The v1.6.1 tag was created without having bumped the version in
`canonicaljson.py`. This means that installing from source at the `v1.6.1` tag will
install the package with the wrong version in its metadata. The 1.6.1 artefacts
uploaded to PyPI did have the correct version; `pip install`ing
directly from the archive resulted in a consistent version number.

Version 1.6.1 released 2022-05-02

* Add type annotations.

Version 1.6.0 released 2022-03-04

* `frozendict` is now an optional dependency; it is no longer required.

Version 1.5.0 released 2021-10-20

* Switch CI from Travis to Github Actions
* Add code to handle frozendict implementations using c-extension
* Add tests for Python 3.10
* Remove outdated Debian packaging code

Version 1.4.0 released 2020-09-03

* Fix producing non-standard JSON for Infinity, -Infinity, and NaN. This could
  cause errors when encoding objects into canonical JSON that previously used to
  work, but were incompatible with JSON implementations in other languages.
* Use UTF-8 to fix ASCII encoding errors when data containing Unicode was
  attempted to be pretty-printed.

Version 1.3.0 released 2020-08-14

* The minimum version of simplejson was bumped to 3.14.0.
* Obsolete workaround for slow encoding of Unicode characters was removed.
* New APIs were added to iteratively encode JSON.

Version 1.2.0 released 2020-07-27

* JSON from the standard library is used automatically on PyPy.
* Support for Python versions which are end-of-lifed was dropped, Python >= 3.5
  is supported and tested in continuous integration.
* An API to configure the underlying JSON library was added (`set_json_library`).

Version 1.1.4 released 2018-05-23

 * Fix error when encoding non-BMP characters on UCS-2 python builds
   (fixes issue #12).

Version 1.1.3 released 2018-04-13

 * Bump depencency on frozendict to >=1.0, to fix conflicts with older
   versions.

Version 1.1.2 released 2018-04-12

 * Fix escaping of control characters U+0000 to U+001F AGAIN, which was STILL
   broken in the previous release

Version 1.1.1 released 2018-04-11

 * Fix escaping of control characters U+0000 to U+001F, which was broken in
   the previous release

Version 1.1.0 released 2018-04-06

 * Significant performance improvements
   ([\#7](https://github.com/matrix-org/python-canonicaljson/pull/7),
   [\#8](https://github.com/matrix-org/python-canonicaljson/pull/8),
   [\#9](https://github.com/matrix-org/python-canonicaljson/pull/9))

Version 1.0.0 released 2015-08-21

 * Initial release


test_canonicaljson.py
# -*- coding: utf-8 -*-
#
# Copyright 2015 OpenMarket Ltd
# Copyright 2018 New Vector Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from math import inf, nan

from canonicaljson import (
    encode_canonical_json,
    encode_pretty_printed_json,
    frozendict_type,
    iterencode_canonical_json,
    iterencode_pretty_printed_json,
    set_json_library,
)

import unittest
from unittest import mock


class TestCanonicalJson(unittest.TestCase):
    def test_encode_canonical(self) -> None:
        self.assertEqual(encode_canonical_json({}), b"{}")

        # ctrl-chars should be encoded.
        self.assertEqual(
            encode_canonical_json(u"text\u0003\r\n"),
            b'"text\\u0003\\r\\n"',
        )

        # quotes and backslashes should be escaped.
        self.assertEqual(
            encode_canonical_json(r'"\ test'),
            b'"\\"\\\\ test"',
        )

        # non-ascii should come out utf8-encoded.
        self.assertEqual(
            encode_canonical_json({u"la merde amusée": u"💩"}),
            b'{"la merde amus\xc3\xa9e":"\xF0\x9F\x92\xA9"}',
        )

        # so should U+2028 and U+2029
        self.assertEqual(
            encode_canonical_json({u"spaces": u"\u2028 \u2029"}),
            b'{"spaces":"\xe2\x80\xa8 \xe2\x80\xa9"}',
        )

        # but we need to watch out for 'u1234' after backslash, which should
        # get encoded to an escaped backslash, followed by u1234
        self.assertEqual(
            encode_canonical_json(u"\\u1234"),
            b'"\\\\u1234"',
        )

        # Iteratively encoding should work.
        self.assertEqual(list(iterencode_canonical_json({})), [b"{}"])

    def test_ascii(self) -> None:
        """
        Ensure the proper ASCII characters are escaped.

        See https://matrix.org/docs/spec/appendices#grammar.
        """
        # Some characters go to their common shorthands.
        escaped = {
            0x08: b'"\\b"',
            0x09: b'"\\t"',
            0x0A: b'"\\n"',
            0x0C: b'"\\f"',
            0x0D: b'"\\r"',
            0x22: b'"\\""',
            0x5C: b'"\\\\"',
        }
        for c, expected in escaped.items():
            self.assertEqual(encode_canonical_json(chr(c)), expected)

        # Others go to the \uXXXX.
        hex_escaped = list(range(0x08)) + [0x0B] + list(range(0x0E, 0x20))
        for c in hex_escaped:
            self.assertEqual(encode_canonical_json(chr(c)), b'"\\u00%02x"' % (c,))

        # And other characters are passed unescaped.
        unescaped = [0x20, 0x21] + list(range(0x23, 0x5C)) + list(range(0x5D, 0x7E))
        for c in unescaped:
            s = chr(c)
            self.assertEqual(encode_canonical_json(s), b'"' + s.encode("ascii") + b'"')

    def test_encode_pretty_printed(self) -> None:
        self.assertEqual(encode_pretty_printed_json({}), b"{}")
        self.assertEqual(list(iterencode_pretty_printed_json({})), [b"{}"])

        # non-ascii should come out utf8-encoded.
        self.assertEqual(
            encode_pretty_printed_json({u"la merde amusée": u"💩"}),
            b'{\n    "la merde amus\xc3\xa9e": "\xF0\x9F\x92\xA9"\n}',
        )

    @unittest.skipIf(
        frozendict_type is None,
        "If `frozendict` is not available, skip test",
    )
    def test_frozen_dict(self) -> None:
        # For mypy's benefit:
        assert frozendict_type is not None
        self.assertEqual(
            encode_canonical_json(frozendict_type({"a": 1})),
            b'{"a":1}',
        )
        self.assertEqual(
            encode_pretty_printed_json(frozendict_type({"a": 1})),
            b'{\n    "a": 1\n}',
        )

    def test_unknown_type(self) -> None:
        class Unknown(object):
            pass

        unknown_object = Unknown()
        with self.assertRaises(Exception):
            encode_canonical_json(unknown_object)

        with self.assertRaises(Exception):
            encode_pretty_printed_json(unknown_object)

    def test_invalid_float_values(self) -> None:
        """Infinity/-Infinity/NaN are not allowed in canonicaljson."""

        with self.assertRaises(ValueError):
            encode_canonical_json(inf)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(inf)

        with self.assertRaises(ValueError):
            encode_canonical_json(-inf)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(-inf)

        with self.assertRaises(ValueError):
            encode_canonical_json(nan)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(nan)

    def test_set_json(self) -> None:
        """Ensure that changing the underlying JSON implementation works."""
        mock_json = mock.Mock(spec=["JSONEncoder"])
        mock_json.JSONEncoder.return_value.encode.return_value = "sentinel"
        try:
            set_json_library(mock_json)
            self.assertEqual(encode_canonical_json({}), b"sentinel")
        finally:
            # Reset the JSON library to whatever was originally set.
            from canonicaljson import json  # type: ignore[attr-defined]

            set_json_library(json)


tests/test_canonicaljson.py
# Copyright 2015 OpenMarket Ltd
# Copyright 2018 New Vector Ltd
# Copyright 2022 The Matrix.org Foundation C.I.C.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import unittest
from math import inf, nan
from unittest.mock import Mock

from canonicaljson import (
    encode_canonical_json,
    encode_pretty_printed_json,
    iterencode_canonical_json,
    iterencode_pretty_printed_json,
    register_preserialisation_callback,
)


class TestCanonicalJson(unittest.TestCase):
    def test_encode_canonical(self) -> None:
        self.assertEqual(encode_canonical_json({}), b"{}")

        # ctrl-chars should be encoded.
        self.assertEqual(
            encode_canonical_json("text\u0003\r\n"),
            b'"text\\u0003\\r\\n"',
        )

        # quotes and backslashes should be escaped.
        self.assertEqual(
            encode_canonical_json(r'"\ test'),
            b'"\\"\\\\ test"',
        )

        # non-ascii should come out utf8-encoded.
        self.assertEqual(
            encode_canonical_json({"la merde amusée": "💩"}),
            b'{"la merde amus\xc3\xa9e":"\xF0\x9F\x92\xA9"}',
        )

        # so should U+2028 and U+2029
        self.assertEqual(
            encode_canonical_json({"spaces": "\u2028 \u2029"}),
            b'{"spaces":"\xe2\x80\xa8 \xe2\x80\xa9"}',
        )

        # but we need to watch out for 'u1234' after backslash, which should
        # get encoded to an escaped backslash, followed by u1234
        self.assertEqual(
            encode_canonical_json("\\u1234"),
            b'"\\\\u1234"',
        )

        # Iteratively encoding should work.
        self.assertEqual(list(iterencode_canonical_json({})), [b"{}"])

    def test_ascii(self) -> None:
        """
        Ensure the proper ASCII characters are escaped.

        See https://matrix.org/docs/spec/appendices#grammar.
        """
        # Some characters go to their common shorthands.
        escaped = {
            0x08: b'"\\b"',
            0x09: b'"\\t"',
            0x0A: b'"\\n"',
            0x0C: b'"\\f"',
            0x0D: b'"\\r"',
            0x22: b'"\\""',
            0x5C: b'"\\\\"',
        }
        for c, expected in escaped.items():
            self.assertEqual(encode_canonical_json(chr(c)), expected)

        # Others go to the \uXXXX.
        hex_escaped = list(range(0x08)) + [0x0B] + list(range(0x0E, 0x20))
        for c in hex_escaped:
            self.assertEqual(encode_canonical_json(chr(c)), b'"\\u00%02x"' % (c,))

        # And other characters are passed unescaped.
        unescaped = [0x20, 0x21] + list(range(0x23, 0x5C)) + list(range(0x5D, 0x7E))
        for c in unescaped:
            s = chr(c)
            self.assertEqual(encode_canonical_json(s), b'"' + s.encode("ascii") + b'"')

    def test_encode_pretty_printed(self) -> None:
        self.assertEqual(encode_pretty_printed_json({}), b"{}")
        self.assertEqual(list(iterencode_pretty_printed_json({})), [b"{}"])

        # non-ascii should come out utf8-encoded.
        self.assertEqual(
            encode_pretty_printed_json({"la merde amusée": "💩"}),
            b'{\n    "la merde amus\xc3\xa9e": "\xF0\x9F\x92\xA9"\n}',
        )

    def test_unknown_type(self) -> None:
        class Unknown(object):
            pass

        unknown_object = Unknown()
        with self.assertRaises(Exception):
            encode_canonical_json(unknown_object)

        with self.assertRaises(Exception):
            encode_pretty_printed_json(unknown_object)

    def test_invalid_float_values(self) -> None:
        """Infinity/-Infinity/NaN are not allowed in canonicaljson."""

        with self.assertRaises(ValueError):
            encode_canonical_json(inf)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(inf)

        with self.assertRaises(ValueError):
            encode_canonical_json(-inf)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(-inf)

        with self.assertRaises(ValueError):
            encode_canonical_json(nan)

        with self.assertRaises(ValueError):
            encode_pretty_printed_json(nan)

    def test_encode_unknown_class_raises(self) -> None:
        class C:
            pass

        with self.assertRaises(Exception):
            encode_canonical_json(C())

    def test_preserialisation_callback(self) -> None:
        class C:
            pass

        # Naughty: this alters the global state of the module. However this
        # `C` class is limited to this test only, so this shouldn't affect
        # other types and other tests.
        register_preserialisation_callback(C, lambda c: "I am a C instance")

        result = encode_canonical_json(C())
        self.assertEqual(result, b'"I am a C instance"')

    def test_cannot_register_preserialisation_callback_for_object(self) -> None:
        with self.assertRaises(Exception):
            register_preserialisation_callback(
                object, lambda c: "shouldn't be able to do this"
            )

    def test_most_recent_preserialisation_callback_called(self) -> None:
        class C:
            pass

        callback1 = Mock(return_value="callback 1 was called")
        callback2 = Mock(return_value="callback 2 was called")

        # Naughty: this alters the global state of the module. However this
        # `C` class is limited to this test only, so this shouldn't affect
        # other types and other tests.
        register_preserialisation_callback(C, callback1)
        register_preserialisation_callback(C, callback2)

        encode_canonical_json(C())

        callback1.assert_not_called()
        callback2.assert_called_once()


benchmark.py
#!/usr/bin/env python
#
# quick benchmark script which loads a bunch of data from an input file, then
# times how long it takes to serialise it

from __future__ import print_function
import json as json
import sys
import timeit

from frozendict import frozendict

import canonicaljson

# try to measure cpu usage time rather than wallclock time, if possible.
try:
    import resource
    RUSAGE_THREAD = 1

    # If the system doesn't support RUSAGE_THREAD then this should throw an
    # exception.
    resource.getrusage(RUSAGE_THREAD)

    def timer():
        r = resource.getrusage(RUSAGE_THREAD)
        return r.ru_utime + r.ru_stime

except Exception:
    timer = timeit.default_timer


if __name__ == "__main__":
    with open(sys.argv[1]) as f:
        small_objs = [frozendict(json.loads(line)) for line in f]

    def serialize():
        for o in small_objs:
            canonicaljson.encode_canonical_json(o)

    time = timeit.timeit(serialize, number=10, timer=timer)

    print("%f seconds" % time)


src/canonicaljson/__init__.py
# Copyright 2014 OpenMarket Ltd
# Copyright 2018 New Vector Ltd
# Copyright 2022 The Matrix.org Foundation C.I.C.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import functools
import json
from typing import Callable, Generator, Type, TypeVar

__version__ = "2.0.0"


@functools.singledispatch
def _preprocess_for_serialisation(obj: object) -> object:  # pragma: no cover
    """Transform an `obj` into something the JSON library knows how to encode.

    This is only called for types that the JSON library does not recognise.
    """
    raise TypeError(
        "Object of type %s is not JSON serializable" % obj.__class__.__name__
    )


T = TypeVar("T")


def register_preserialisation_callback(
    data_type: Type[T], callback: Callable[[T], object]
) -> None:
    """
    Register a `callback` to preprocess `data_type` objects unknown to the JSON encoder.

    When canonicaljson encodes an object `x` at runtime that its JSON library does not
    know how to encode, it will
      - select a `callback`,
      - compute `y = callback(x)`, then
      - JSON-encode `y` and return the result.

    The `callback` should return an object that is JSON-serialisable by the stdlib
    json module.

    If this is called multiple times with the same `data_type`, the most recently
    registered callback is used when serialising that `data_type`.
    """
    if data_type is object:
        raise ValueError("Cannot register callback for the `object` type")
    _preprocess_for_serialisation.register(data_type, callback)


# Declare these once for re-use.
_canonical_encoder = json.JSONEncoder(
    ensure_ascii=False,
    allow_nan=False,
    separators=(",", ":"),
    sort_keys=True,
    default=_preprocess_for_serialisation,
)
_pretty_encoder = json.JSONEncoder(
    ensure_ascii=False,
    allow_nan=False,
    indent=4,
    sort_keys=True,
    default=_preprocess_for_serialisation,
)


def encode_canonical_json(data: object) -> bytes:
    """Encodes the given `data` as a UTF-8 canonical JSON bytestring.

    This encoding is the shortest possible. Dictionary keys are
    lexicographically sorted by unicode code point.
    """
    s = _canonical_encoder.encode(data)
    return s.encode("utf-8")


def iterencode_canonical_json(data: object) -> Generator[bytes, None, None]:
    """Iteratively encodes the given `data` as a UTF-8 canonical JSON bytestring.

    This yields one or more bytestrings; concatenating them all together yields the
    full encoding of `data`. Building up the encoding gradually in this way allows us to
    encode large pieces of `data` without blocking other tasks.

    This encoding is the shortest possible. Dictionary keys are
    lexicographically sorted by unicode code point.
    """
    for chunk in _canonical_encoder.iterencode(data):
        yield chunk.encode("utf-8")


def encode_pretty_printed_json(data: object) -> bytes:
    """Encodes the given `data` as a UTF-8 human-readable JSON bytestring."""

    return _pretty_encoder.encode(data).encode("utf-8")


def iterencode_pretty_printed_json(data: object) -> Generator[bytes, None, None]:
    """Iteratively encodes the given `data` as a UTF-8 human-readable JSON bytestring.

    This yields one or more bytestrings; concatenating them all together yields the
    full encoding of `data`. Building up the encoding gradually in this way allows us to
    encode large pieces of `data` without blocking other tasks.
    """
    for chunk in _pretty_encoder.iterencode(data):
        yield chunk.encode("utf-8")


canonicaljson.py
# -*- coding: utf-8 -*-

# Copyright 2014 OpenMarket Ltd
# Copyright 2018 New Vector Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import platform
from typing import Any, Generator, Optional, Type

try:
    from typing import Protocol
except ImportError:  # pragma: no cover
    from typing_extensions import Protocol  # type: ignore[misc]

frozendict_type: Optional[Type[Any]]
try:
    from frozendict import frozendict as frozendict_type
except ImportError:
    frozendict_type = None  # pragma: no cover

__version__ = "1.6.2"


def _default(obj: object) -> object:  # pragma: no cover
    if type(obj) is frozendict_type:
        # If frozendict is available and used, cast `obj` into a dict
        return dict(obj)  # type: ignore[call-overload]
    raise TypeError(
        "Object of type %s is not JSON serializable" % obj.__class__.__name__
    )


class Encoder(Protocol):  # pragma: no cover
    def encode(self, data: object) -> str:
        pass

    def iterencode(self, data: object) -> Generator[str, None, None]:
        pass

    def __call__(self, *args: Any, **kwargs: Any) -> "Encoder":
        pass


class JsonLibrary(Protocol):
    JSONEncoder: Encoder


# Declare these in the module scope, but they get configured in
# set_json_library.
_canonical_encoder: Encoder = None  # type: ignore[assignment]
_pretty_encoder: Encoder = None  # type: ignore[assignment]


def set_json_library(json_lib: JsonLibrary) -> None:
    """
    Set the underlying JSON library that canonicaljson uses to json_lib.

    Params:
        json_lib: The module to use for JSON encoding. Must have a
            `JSONEncoder` property.
    """
    global _canonical_encoder
    _canonical_encoder = json_lib.JSONEncoder(
        ensure_ascii=False,
        allow_nan=False,
        separators=(",", ":"),
        sort_keys=True,
        default=_default,
    )

    global _pretty_encoder
    _pretty_encoder = json_lib.JSONEncoder(
        ensure_ascii=False,
        allow_nan=False,
        indent=4,
        sort_keys=True,
        default=_default,
    )


def encode_canonical_json(data: object) -> bytes:
    """Encodes the given `data` as a UTF-8 canonical JSON bytestring.

    This encoding is the shortest possible. Dictionary keys are
    lexicographically sorted by unicode code point.
    """
    s = _canonical_encoder.encode(data)
    return s.encode("utf-8")


def iterencode_canonical_json(data: object) -> Generator[bytes, None, None]:
    """Iteratively encodes the given `data` as a UTF-8 canonical JSON bytestring.

    This yields one or more bytestrings; concatenating them all together yields the
    full encoding of `data`. Building up the encoding gradually in this way allows us to
    encode large pieces of `data` without blocking other tasks.

    This encoding is the shortest possible. Dictionary keys are
    lexicographically sorted by unicode code point.
    """
    for chunk in _canonical_encoder.iterencode(data):
        yield chunk.encode("utf-8")


def encode_pretty_printed_json(data: object) -> bytes:
    """Encodes the given `data` as a UTF-8 human-readable JSON bytestring."""

    return _pretty_encoder.encode(data).encode("utf-8")


def iterencode_pretty_printed_json(data: object) -> Generator[bytes, None, None]:
    """Iteratively encodes the given `data` as a UTF-8 human-readable JSON bytestring.

    This yields one or more bytestrings; concatenating them all together yields the
    full encoding of `data`. Building up the encoding gradually in this way allows us to
    encode large pieces of `data` without blocking other tasks.
    """
    for chunk in _pretty_encoder.iterencode(data):
        yield chunk.encode("utf-8")


if platform.python_implementation() == "PyPy":  # pragma: no cover
    # pypy ships with an optimised JSON encoder/decoder that is faster than
    # simplejson's C extension.
    import json
else:  # pragma: no cover
    # using simplejson rather than regular json on CPython for backwards
    # compatibility (simplejson on Python 3.5 handles parsing of bytes while
    # the standard library json does not).
    #
    # Note that it seems performance is on par or better using json from the
    # standard library as of Python 3.7.
    import simplejson as json  # type: ignore[no-redef]

# Set the JSON library to the backwards compatible version.
set_json_library(json)
